swiperefresh/src/main/java/com/google/accompanist/swiperefresh/SwipeRefresh.kt

```diff
@@ -0,0 +1,336 @@
+ * swipe gesture.
+ *
+ * Apps should provide a [onRefresh] block to be notified each time a swipe to refresh gesture
+ * is completed. That block is responsible for updating the [state] as appropriately,
+ * typically by setting [SwipeRefreshState.isRefreshing] to `true` once a 'refresh' has been
+ * started. Once a refresh has completed, the app should then set
+ * [SwipeRefreshState.isRefreshing] to `false`.
+ *
+ * If an app wishes to show just the progress animation, outside of a swipe refresh, it can
+ * set [SwipeRefreshState.isRefreshing] as required.
+ *
+ * @sample com.google.accompanist.sample.swiperefresh.SwipeRefreshSample
+ *
+ * @param state the state object to be used to control or observe the [SwipeRefresh] state.
+ * @param onRefresh Lambda which is invoked when a swipe to refresh gesture is completed.
+ * @param modifier the modifier to apply to this layout.
+ * @param swipeEnabled Whether the the layout should react to swipe gestures or not.
+ * @param indicator the indicator that represents the current state. By default this
+ * will be a [SwipeRefreshIndicator].
+ * @param content The content containing a scroll composable.
+ */
+@Composable
+fun SwipeRefresh(
+    state: SwipeRefreshState,
+    onRefresh: () -> Unit,
+    modifier: Modifier = Modifier,
+    swipeEnabled: Boolean = true,
+    indicator: @Composable SwipeRefreshIndicatorScope.() -> Unit = {
+        SwipeRefreshIndicator(
+            isRefreshing = isRefreshing,
+            offset = indicatorOffset,
+            triggerOffset = indicatorTriggerOffset,
+        )
+    },
+    content: @Composable () -> Unit,
+) {
+    val coroutineScope = rememberCoroutineScope()
+    val updatedOnRefresh = rememberUpdatedState(onRefresh)
+
+    // Our LaunchedEffect, which animates the indicator to an appropriate resting position
+    LaunchedEffect(state.isSwipeInProgress, state.isRefreshing) {
+        if (!state.isSwipeInProgress) {
+            // If there's not a swipe in progress, rest the indicator at an appropriate position
+            state.animateBackToRest()
+        }
+    }
+
+    // Our nested scroll connection, which updates our state.
+    val nestedScrollConnection = remember(state, coroutineScope) {
+        SwipeRefreshNestedScrollConnection(state, coroutineScope) {
+            // On refresh, re-dispatch to the update onRefresh block
+            updatedOnRefresh.value.invoke()
+        }
+    }.apply {
+        this.enabled = swipeEnabled
+    }
+
+    val indicatorContentScope = remember(state) { SwipeRefreshIndicatorScopeImpl(state) }
+
+    Layout(
+        content = {
+            Box(Modifier.layoutId(LayoutContentTag)) { content() }
+            Box(Modifier.layoutId(LayoutIndicatorTag)) { indicatorContentScope.indicator() }
+        },
+        modifier = modifier
+            .nestedScroll(connection = nestedScrollConnection)
+            .clipToBounds()
```

Comment:
Question: Do you really need to clip? It's ok I guess, but you might want to leave this for a scroll-container implementations (Modifier.verticalScroll, LazyColumn) and they already do it by default.

You might save a render node/ a view here + allow developers to not to clip if they have their own components that does the job slightly different.

Up to you though.

Reply:
I think we need to allow apps to clip, as without it the indicator will drawn when it's 'off-screen'. Apps supplying their own `clipToBounds()` modifier is fine though I think. I'll add a comment for this.